app/src/features/surveys/view/survey-animals/SurveyAnimalsTableActions.tsx

```diff
@@ -53,22 +55,25 @@ const SurveyAnimalsTableActions = (props: ITableActionsMenuProps) => {
         MenuListProps={{
           'aria-labelledby': 'basic-button'
         }}>
-        <MenuItem
-          onClick={() => {
-            handleClose();
-            props.onAddDevice(props.critter_id);
-          }}
-          data-testid="animal-table-row-add-device">
-          <ListItemIcon>
-            <Icon path={mdiPlus} size={1} />
-          </ListItemIcon>
-          <Typography variant="inherit">Add Telemetry Device</Typography>
-        </MenuItem>
-        {props.devices?.length ? (
+        {props.onAddDevice ? (
           <MenuItem
             onClick={() => {
               handleClose();
-              props.onEditDevice(props.critter_id);
+              props.onAddDevice?.(props.critter_id);
+            }}
+            data-testid="animal-table-row-add-device">
+            <ListItemIcon>
+              <Icon path={mdiPlus} size={1} />
+            </ListItemIcon>
+            <Typography variant="inherit">Add Telemetry Device</Typography>
+          </MenuItem>
+        ) : null}
+
+        {props.devices?.length && props.onEditDevice ? (
+          <MenuItem
+            onClick={() => {
+              handleClose();
+              props.onEditDevice?.(props.critter_id);
```

Comment:
Getting error: Cannot invoke an object which is possibly 'undefined'.

Which is strange cause I am checking on line 72 if its defined

Reply:
Its because `onEditDevice` is being executed inside a callback function, and typescript can't be certain if `onEditDevice` will always be defined (since the callback is going to be executed in the future, and its possible for the function to become undefined before it gets run).

This post talks about it: https://stackoverflow.com/questions/68371833/cannot-invoke-an-object-which-is-possibly-undefined-even-after-ensuring-it

Reply:
I vaguely remember having a similar issue with codes somewhere. We'd check that codes wasn't null, but we'd still get complaints about codes possibly being null, because it was being referenced inside some nested child function or something.